### ISSUE TITLE
ActiveRelationTrait::filterByModels prefix issue with composite key

### ISSUE THREAD

**Issue by airmoi** [yiisoft/yii2#10201]
When using a composite key in a relation that include a join, the filterByModels method calls the prefixKeyColumns method that add relate table alias to the attribute.

Then it generates the values array indexed by attributes names without including the table alias added previously.

This cause a missmatch between attributes and values when the "in" condition is builded and produce a null query.

Here is a sample relation that produce the issue : 
```php
public function getItemsWithCategories()   {
     return $this->hasOne(Items::className(), [
            'a_id' => 'a_id',
            'b_id' => 'b_id'
        ])
        ->joinWith('categories');
    }
```

Using this relation, filterByModels calls the andWhere method with this parameters : 
```php
$attributes = [0 => 'items.a_id', 1 =>' items.b.id' ] ;
$values = [[ 'a_id' => '1', 'b_id' => '5' ] , ..];
```

Final query looks like this : 
```sql
SELECT * FROM items LEFT JOIN categories ON items.id_category = categories.id WHERE items.a_id IS NULL and items.id_b IS NULL
```

**Comment by cebe** [comment]
could you create a unit test to show how to reproduce it?

**Comment by airmoi** [comment]
I'm not comfortable with unit test but i'll try to provide you something

**Comment by airmoi** [comment]
Ok here's a test unit case (based on framework unit test)

Add these 2 relations in data/ar/OrderItem.php
```php
    public function getOrderItemCompositeWithJoin()
    {
        return $this->hasOne(OrderItem::className(), ['item_id' => 'item_id', 'order_id' => 'order_id' ])
                ->joinWith('item');
    }

    public function getOrderItemCompositeNoJoin()
    {
        return $this->hasOne(OrderItem::className(), ['item_id' => 'item_id', 'order_id' => 'order_id' ]);
    }
```

Then add the folowing test case in framework/db/ActiveRecordTest.php
```php
public function testFindCompositeWithJoin()
    {
        /* @var $orderItem OrderItem */
        $orderItem = OrderItem::findOne([1, 1]);
        
        $orderItemNoJoin = $orderItem->orderItemCompositeNoJoin;
        $this->assertInstanceOf('yiiunit\data\ar\OrderItem', $orderItemNoJoin);
        
        $orderItemWithJoin = $orderItem->orderItemCompositeWithJoin;
        $this->assertInstanceOf('yiiunit\data\ar\OrderItem', $orderItemWithJoin);
    }
```

Running the test will pass the first asset while the second will fail

**Comment by cebe** [comment]
Awesome, thanks! :)

**Comment by PaulVanSchayck** [comment]
I also stumbled upon this bug. Rather rare case with composite key and join with IN condition.

Would adding this to `QueryBuilder->buildCompositeInCondition()` be too simplistic?
```
if ( ($pos = strpos($column, '.')) !== false ) {
    $column = substr($column, $pos+1);
}
```

Or adding something like this to `ActiveRelationTrait->filterByModels()`
```
            // composite keys
            foreach ($models as $model) {
                $v = [];
                foreach ($this->link as $attribute => $link) {
                    $t = $this->prefixKeyColumns([$attribute]);
                    $v[$t[0]] = $model[$link];
                }
                $values[] = $v;
            }
```
@airmoi Did you find a generic workaround?

**Comment by robsch** [comment]
I've also got this problem. Does anyone know a workaround (now)?

**Comment by juffin-halli** [comment]
Same issue:
```
$attributes = $this->prefixKeyColumns($attributes);
...
// composite keys
foreach ($models as $model) {
    $v = [];
    foreach ($this->link as $attribute => $link) {
        $v[$attribute] = $model[$link];
    }
    $values[] = $v;
}
...
$this->andWhere(['in', $attributes, array_unique($values, SORT_REGULAR)]);
```
I think $attribute should be mapped to prefixed attribute.

**Comment by cebe** [comment]
hm...  is this a duplicate of #9895/#9047 ?

**Comment by PaulVanSchayck** [comment]
@cebe I haven't tested it. But looking at the (tested) solution provided [by me above](#issuecomment-164589088), and at the solution provided in PR #9895 this looks like the same issue. Or at least the same solution. 

@airmoi provided also a [unit test above](#issuecomment-158473854)

**Comment by cebe** [comment]
yep, thanks. I am able to reproduce it with his test.

**Comment by cebe** [comment]
Fixed by 52e23e0, thanks @PaulVanSchayck, @airmoi and @joe-meyer!
